futures/src/transport.rs

```diff
@@ -160,32 +169,44 @@ impl<T> AMQPTransport<T>
 
     if let Some(frame) = value {
       trace!("upstream poll gave frame: {:?}", frame);
-      self.conn.handle_frame(frame);
+      if let Err(e) = self.conn.handle_frame(frame) {
+        let err = format!("failed to handle frame: {:?}", e);
+        return Err(io::Error::new(io::ErrorKind::Other, err));
+      }
       self.send_frames();
-      self.upstream.poll_complete();
       Ok(Async::Ready(Some(())))
     } else {
       error!("upstream poll gave Ready(None)");
       Ok(Async::Ready(None))
     }
   }
 
-  pub fn send_and_handle_frames(&mut self) {
+  pub fn send_and_handle_frames(&mut self) -> Poll<Option<()>, io::Error> {
     self.send_frames();
-    self.handle_frames();
+    self.handle_frames()
   }
 
   pub fn send_frames(&mut self) {
     //FIXME: find a way to use a future here
     while let Some(f) = self.conn.next_frame() {
-      self.upstream.start_send(f);
-      self.upstream.poll_complete();
+      if let Err(e) = self.send_frame(f) {
+        error!("Failed to send frame: {:?}", e);
+        break;
+      }
     }
-    //self.upstream.poll_complete();
   }
 
-  pub fn handle_frames(&mut self) {
-    while let Ok(Async::Ready(Some(_))) = self.poll() {}
+  fn send_frame(&mut self, frame: Frame) -> Poll<(), io::Error> {
+      self.start_send(frame).and_then(|_| self.poll_complete())
```

Comment:
it might be a bit slow to do a poll_complete after each send, maybe there's a way to optimize that